Add enrich policy list API

hub-cap opened the pull request:
This commit wires up the Rest calls and Transport calls for listing all
enrich policies, as well  as tests and rest spec additions.

elasticmachine commented:
Pinging @elastic/es-core-features

hub-cap commented:
@elasticmachine run elasticsearch-ci/1

hub-cap commented:
@elasticmachine run elasticsearch-ci/bwc

martijnvg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/action/ListEnrichPolicyAction.java:
```diff
@@ -0,0 +1,105 @@
+            {
+                builder.startArray("policies");
+                {
+                    for (Map.Entry<String, EnrichPolicy> entry : policyMap.entrySet()) {
```
We iterate here directly from a HashMap. The ordering of the entry set is not deterministic, so different api invocations can result in policies being returned in a different order. 

In order to fix that we should make a copy here using a `TreeMap`:

```
TreeMap<String, EnrichPolicy> policyMap = new TreeMap<>(this.policyMap);
for (Map.Entry<String, EnrichPolicy> entry : policyMap.entrySet()) {
```

hub-cap replied:
I moved this into the constructor since the input is a map and the internal state of the object is a list.

martijnvg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/EnrichPolicy.java:
```diff
@@ -43,16 +44,18 @@
     );
 
     static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), NAME);
```
I think this should be an optional parameter. The put transport action should set the name as it can be derived from the put policy request class. We only want to make setting the name in the url path required.

hub-cap replied:
Removed, now in `NamedPolicy`

jakelandis commented:
LGTM (pending Martijn's comments)

hub-cap commented:
ok, readdressed all comments and worked out the `EnrichPolicy.NamedPolicy`, which is a representation used currently for the list operations. This allows us to inject an additional field, name, into the enrich policy's output and validate it properly in tests. This might be used in other places that require a name and a policy, such as PUT, and GET.

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/rest/RestListEnrichPolicyAction.java:
```diff
@@ -0,0 +1,35 @@
+
+    public RestListEnrichPolicyAction(final Settings settings, final RestController controller) {
+        super(settings);
+        controller.registerHandler(RestRequest.Method.GET, "/_enrich/policy", this);
```
plural? `/_enrich/policies`?

hub-cap replied:
if w go with this then i suggest we change all of the URIs, which i like, but I assumed we did not want to go against for now. This is how I view it. You are enacting on a single or multiple policies at any time. So, `/policies` gives you the list, and `/policies/foo` gives your single foo policy back, but the resource is the same, you are just saying "give me a singlular one of these policies". Otherwise I think we should keep it the same as `policy` across the board. I prefer the former, but its possible we dont want to go that route until more actual API and versioning conversations happen.

martijnvg replied:
ok :) Lets leave this for now and decide whether we want this in a later stage of development.

Btw I just check get mappings api (returns a hash and not a list) and there both singular and plural is supported. But it looks like most apis don't mix plural with singular nouns when it comes returning a single item or multiple (get pipeline, get repository, get tasks etc). Anyhow, language wise (i'm not a native speaker) plural made sense to me.

hub-cap replied:
IMHO plural makes sesne for all cases. It reads easier to me, a list of policies `GET /_enrich/policies`, a single policy from the list of policies `GET /_enrich/policies/foo`
